### PR TITLE
Fetch optimization

### LINKED ISSUE #274: west update: don't fetch remotes if revision is a SHA available locally
As a speed optimization and to work better offline. Leave the current behavior in place under a config option and/or command line flag (like -n / --network).

### PR DESCRIPTION
**Pull request by mbolivar**
Depends on #290, fixes #274

```
By default, skip fetches of revisions that resolve to SHAs or tag
names that are already available locally.

The user can override by using west update -f always.
```

CC: @pabigot

### CODE REVIEW

**Comment by mbolivar** [comment]
First commit not in #290 also fixes: #191

**Comment by mbolivar** [comment]
@pabigot @carlesc on my Linux VM, this takes `west update` from ~5 sec down to ~0.5 sec in the downstream nrf repository. On mainline zephyr, 8 sec becomes ~0.8 sec.

Of course mileage will vary depending on network conditions, file system performance, etc.

**Comment by carlescufi** [comment]
Please rebase now that #290 is merged

**Comment by carlescufi** [comment]
> @pabigot @carlescufi on my Linux VM, this takes `west update` from ~5 sec down to ~0.5 sec in the downstream nrf repository. On mainline zephyr, 8 sec becomes ~0.8 sec.
> 
> Of course mileage will vary depending on network conditions, file system performance, etc.

Those are really good improvements! Since the bulk of west update invocations end up fetching just a couple of repos this will have a big impact on mostly everyone. Nice job!

**Comment by mbolivar** [comment]
> Please rebase now that #290 is merged

Done, thanks for merging

File: src/west/commands/project.py [hunk 1]
```diff
@@ -589,7 +589,18 @@ def do_add_parser(self, parser_adder):
         parser = self._parser(parser_adder)
 
         group = parser.add_argument_group(
-            title='checked out branches',
+            title='fetching behavior',
+            description='By default, west update tries to avoid fetching.')
+        group.add_argument('-f', '--fetch', dest='fetch_strategy',
```
Comment (carlescufi) [inline src/west/commands/project.py]: Should this be a configuration option too? I think so

**Comment by carlescufi** [comment]
I tested this locally and it is pretty amazing :) Thanks @mbolivar for this, will really improve everybody's experience using `west update`.

**Comment by mbolivar** [comment]
> I tested this locally and it is pretty amazing :) Thanks @mbolivar for this, will really improve everybody's experience using `west update`.

Awesome, thanks! I'm glad it's working for you :).

I've added a new `update.fetch` option which takes the same values as the `--fetch` option, so hopefully this is good to go.

**Pull request merged**